lib/trento/domain/sap_system/sap_system.ex

```diff
@@ -290,6 +297,188 @@ defmodule Trento.Domain.SapSystem do
+        sap_system,
+        command,
+        deregistered_instance_role
+      )
+    end)
+    |> Multi.execute(fn sap_system ->
+      maybe_emit_sap_system_deregistered_event(sap_system, command)
+    end)
+  end
+
+  # Deregister a database instance and emit a DatabaseInstanceDeregistered
+  # also potentially emit ApplicationInstanceDeregistered, SapSystemDeregistered
+  # and DatabaseDeregistered events
+  def execute(
+        %SapSystem{sap_system_id: sap_system_id, database: %Database{instances: instances}} =
+          sap_system,
+        %DeregisterDatabaseInstance{
+          instance_number: instance_number,
+          host_id: host_id,
+          sap_system_id: sap_system_id,
+          deregistered_at: deregistered_at
+        } = command
+      ) do
+    instance = get_instance(instances, host_id, instance_number)
+
+    is_primary_deregistered =
+      case instance do
+        nil -> false
+        %Instance{system_replication: "Primary"} -> true
+        %Instance{system_replication: _} -> false
+      end
+
+    sap_system
+    |> Multi.new()
+    |> Multi.execute(fn _ ->
+      %DatabaseInstanceDeregistered{
+        sap_system_id: sap_system_id,
+        host_id: host_id,
+        instance_number: instance_number,
+        deregistered_at: deregistered_at
+      }
+    end)
+    |> Multi.execute(fn sap_system ->
+      maybe_emit_database_instance_deregistered_events(
+        sap_system,
+        command,
+        is_primary_deregistered
+      )
+    end)
+    |> Multi.execute(fn sap_system ->
+      maybe_emit_application_instance_deregistered_events(sap_system, command)
+    end)
+    |> Multi.execute(fn sap_system ->
+      maybe_emit_sap_system_deregistered_event(sap_system, command)
+    end)
+    |> Multi.execute(fn sap_system ->
+      maybe_emit_database_deregistered_event(sap_system, command)
+    end)
+  end
+
+  def apply(
+        %SapSystem{database: %Database{instances: instances}} = sap_system,
+        %DatabaseInstanceDeregistered{
+          instance_number: instance_number,
+          host_id: host_id
+        }
+      ) do
+    instances =
+      Enum.reject(instances, fn
+        %Instance{instance_number: ^instance_number, host_id: ^host_id} ->
+          true
+
+        _ ->
+          false
+      end)
+
+    %SapSystem{
+      sap_system
+      | database: %Database{
+          instances: instances
+        }
+    }
+  end
+
+  def apply(
+        %SapSystem{application: %Application{instances: instances}} = sap_system,
+        %ApplicationInstanceDeregistered{instance_number: instance_number, host_id: host_id}
+      ) do
+    instances =
+      Enum.reject(instances, fn
+        %Instance{instance_number: ^instance_number, host_id: ^host_id} ->
+          true
+
+        _ ->
+          false
+      end)
+
+    %SapSystem{
+      sap_system
+      | application: %Application{
+          instances: instances
+        }
+    }
+  end
+
+  def apply(
```

Comment:
Do we really need this apply?
Cannot we use the next 2 to set database and application to nil?